user/pcmcia-cs/man/ifuser.8
.\" Copyright (C) 1998 David A. Hinds -- dahinds@users.sourceforge.net
.\" ifuser.8 1.6 2000/06/12 21:24:48
.\"
.TH IFUSER 8 "2000/06/12 21:24:48" "pcmcia-cs"
.SH NAME
ifuser \- identify destinations routed to a particular network interface
.SH SYNOPSIS
.B ifuser
.RB [ -v ]
.I interface
.RI [ "target\ ..." ]
.SH DESCRIPTION
This utility checks to see if any of the listed hosts or network
addresses are routed through the specified interface.  Destinations
may be specified either by IP address or by name.
.PP
The exit code will be 0 if any listed host or network is routed
through this interface, and 1 if the interface is not in use (similar
to 
.BR fuser ).
.SH OPTIONS
.TP
.B \-v
Verbose mode: if enabled, then the names of those destinations routed
to this interface (if any) will be reported on standard output.
.SH AUTHORS
David Hinds \- dahinds@users.sourceforge.net
.br
Regis "HPReg" Duchesne \- regis@via.ecp.fr


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/pcmcia-cs/cardmgr/ifuser.c
/*======================================================================

    This utility checks to see if any of a list of hosts or network
    addresses are routed through a specified interface.  Destinations
    may be specified either by IP address or by name.
    
    usage: ifuser [-v] interface [target ...]

    The exit code is 0 if any host is using the specified interface,
    and 1 if the interface is not in use (just like fuser).
    
    ifuser.c 1.15 2000/06/12 21:33:02

    1998/10/24: Regis "HPReg" Duchesne <regis@via.ecp.fr>
      . Added network names (/etc/networks) management
      . Used u_int32_t instead of u_int
      . Handled a malloc error

    The contents of this file are subject to the Mozilla Public
    License Version 1.1 (the "License"); you may not use this file
    except in compliance with the License. You may obtain a copy of
    the License at http://www.mozilla.org/MPL/

    Software distributed under the License is distributed on an "AS
    IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
    implied. See the License for the specific language governing
    rights and limitations under the License.

    The initial developer of the original code is David A. Hinds
    <dahinds@users.sourceforge.net>.  Portions created by David A. Hinds
    are Copyright (C) 1999 David A. Hinds.  All Rights Reserved.

    Alternatively, the contents of this file may be used under the
    terms of the GNU Public License version 2 (the "GPL"), in which
    case the provisions of the GPL are applicable instead of the
    above.  If you wish to allow the use of your version of this file
    only under the terms of the GPL and not to allow others to use
    your version of this file under the MPL, indicate your decision
    by deleting the provisions above and replace them with the notice
    and other provisions required by the GPL.  If you do not delete
    the provisions above, a recipient may use your version of this
    file under either the MPL or the GPL.
    Portions created by Regis "HPReg" Duchesne are Copyright (C) 1998
    Regis "HPReg" Duchesne.  All Rights Reserved.

======================================================================*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <netdb.h>
#include <errno.h>
#include <ctype.h>
#include <netinet/in.h>
#include <arpa/inet.h>

typedef struct route_t {
    u_int32_t		dest, mask;
    int			match;
    struct route_t	*next;
} route_t;

/*====================================================================*/

static int resolv_name(char *s, u_int32_t *a)
{
    struct in_addr addr;
    struct hostent *hp;
    struct netent *np;
    
    if (inet_aton(s, &addr)) {
	*a = (u_int32_t)ntohl(addr.s_addr);
	return 0;
    }
    np = getnetbyname(s);
    if (np) {
	*a = (u_int32_t)np->n_net;
	return 0;
    }
    hp = gethostbyname(s);
    if (hp) {
	*a = (u_int32_t)ntohl(*(u_int32_t *)hp->h_addr_list[0]);
	return 0;
    }
    return -1;
}

/*====================================================================*/

static void usage(char *s)
{
    fprintf(stderr, "usage: %s [-v] interface [target ...]\n", s);
    exit(1);
}

int main(int argc, char *argv[])
{
    char *dev, s[129], dest[16], mask[16], iface[10];
    route_t *r, *tbl, **tail;
    int i, verbose = 0, busy = 0;
    FILE *f;

    i = 1;
    if (argc < 2) usage(argv[0]);
    if (strcmp(argv[1], "-v") == 0) {
	verbose = 1; i++;
    }
    if ((*argv[i] == '-') || (argc < i+1)) usage(argv[0]);
    dev = argv[i]; i++;
    
    /* Get routing table */
    f = popen("netstat -nr", "r");
    if (f == NULL) {
	fprintf(stderr, "%s: could not get routing table: %s\n",
		argv[0], strerror(errno));
	return 2;
    }
    
    do {
	fgets(s, 128, f);
    } while (!feof(f) && !isdigit(s[0]));
    
    tail = &tbl;
    do {
	r = malloc(sizeof(route_t));
	if (r == NULL) {
	    fprintf(stderr, "%s: out of memory\n", argv[0]);
	    return 2;
	}
	sscanf(s, "%s %*s %s %*s %*s %*s %*s %s", dest, mask, iface);
	resolv_name(dest, &r->dest);
	resolv_name(mask, &r->mask);
	r->match = (strcmp(iface, dev) == 0);
	*tail = r; tail = &(r->next);
    } while (fgets(s, 128, f) != NULL);
    *tail = NULL;
    pclose(f);

    /* Check each host on command line */
    for (; i < argc; i++) {
	u_int32_t a;
	if (resolv_name(argv[i], &a) != 0) {
	    fprintf(stderr, "%s: lookup failed: %s\n",
		    argv[0], argv[i]);
	    continue;
	}

	for (r = tbl; r; r = r->next) {
	    if ((a & r->mask) == r->dest) {
		if (r->match) {
		    if (verbose) {
			if (!busy) printf("%s:", dev);
			printf(" %s", argv[i]);
		    }
		    busy = 1;
		}
		break;
	    }
	}
    }
    
    if (busy && verbose)
	printf("\n");
    return (!busy);
}
